shared/src/question-generators/propositional-logic/normalForms.ts

```diff
@@ -0,0 +1,281 @@
+    },
+    () => {
+      const tmpExpr = expression.copy().invertRandomLiterals(random, 1)
+      return isDNF ? tmpExpr.toDNF() : tmpExpr.toCNF()
+    },
+    () => {
+      const tmpExpr = expression.copy().invertRandomLiterals(random, 1)
+      return random.bool() ? tmpExpr.toDNF() : tmpExpr.toCNF()
+    },
+  ]
+
+  const wrongAnswersStrategy = random.choice(wrongAnswerStrategies)
+  while (answers.length < 3) {
+    newWrongAnswer = wrongAnswersStrategy()
+    newWrongAnswerStr = newWrongAnswer.toString(true)
+    if (
+      correctAnswerTruthTable != newWrongAnswer.getTruthTable().truthTable &&
+      answers.every((answerStr) => answerStr !== newWrongAnswerStr)
+    )
+      answers.push(newWrongAnswerStr)
+  }
+
+  answers.push(correctAnswerStr)
+  random.shuffle(answers)
+
+  const correctAnswerIndex = answers.indexOf(correctAnswerStr)
+
+  return <MultipleChoiceQuestion>{
+    type: "MultipleChoiceQuestion",
+    name: NormalForms.name(lang),
+    path: path,
+    text: t(translations, lang, "choice_question", [
+      expression.toString(true),
+      t(translations, lang, isDNF ? "dnf" : "cnf"),
+    ]),
+    answers: answers.map((str) => `$${str}$`),
+    allowMultiple: true,
+    feedback: minimalMultipleChoiceFeedback({
+      correctAnswerIndex,
+      // todo add feedback depending on selection? how?
+      // feedbackText: getAdditionalFeedbackText(satisfiable, falsifiable, lang),
+    }),
+  }
+}
+
+function getCorrectNormalFormStr(expression: SyntaxTreeNodeType, isDNF: boolean): string {
+  return "$" + (isDNF ? expression.toDNF() : expression.toCNF()).toString(true) + "$"
+}
+
+function getFreeTextFeedbackFunction(
+  lang: Language,
+  expression: SyntaxTreeNodeType,
+  isDNF: boolean,
+): FreeTextFeedbackFunction {
+  return ({ text }) => {
+    try {
+      const answer = PropositionalLogicParser.parse(text)
+      if (isDNF ? answer.isDNF() : answer.isCNF)
+        return {
+          correct: false,
+          correctAnswer: "$" + getCorrectNormalFormStr(expression, isDNF) + "$",
+          feedbackText: t(translations, lang, "freetext_feedback_no_normal_form"),
+        }
+
+      const { variableNames: exprVars, truthTable: exprTruthTable } = expression.getTruthTable()
+      const { variableNames: answerVars, truthTable: answerTruthTable } = answer.getTruthTable()
+
+      if (exprVars !== answerVars)
+        return {
+          correct: false,
+          correctAnswer: getCorrectNormalFormStr(expression, isDNF),
+          feedbackText: t(translations, lang, "freetext_feedback_unknown_variables"),
+        }
+
+      if (exprTruthTable !== answerTruthTable)
+        return {
+          correct: false,
+          correctAnswer: getCorrectNormalFormStr(expression, isDNF),
+          feedbackText: t(translations, lang, "freetext_feedback_not_equivalent"),
+        }
+
+      return {
+        correct: true,
+      }
+    } catch (e) {
```

Comment:
The parser can throw exceptions. The exceptions contain information what they expected and where which could be displayed in the feedback (but isn't at the moment). If the try-catch block only contains the parse-call, `answer` cannot be `const` and needs to be declared outside.

Reply:
I see. I guess I prefer the pattern where the parser is always safe, that is, instead of throwing an exception it returns an error. However, let's not get bogged down with this too much, do whatever you feel is best.

Reply:
I see. The reason I use an exception is that the parser is recursive and so I don't need to check the return value every time. The exception just bubbles all the way to the top. However, I changed it the following way: There is a helper function that makes the parser call, catches the exception and then returns either the syntax tree or the parser error. This has also the benefit that the generator does not have to handle the exception correctly. At the moment it catches every exception that might occur, but actually only the ParserError should be handled and everything else should be rethrown. If anything can happen anyway, I have not checked.